vdr/src/client/client.rs

```diff
@@ -302,64 +302,214 @@ pub mod test {
+            signature: RwLock::new(None),
+            hash: None,
+        };
+        let signer = basic_signer();
+        let sign_bytes = transaction.get_signing_bytes().unwrap();
+        let signature = signer.sign(&sign_bytes, TRUSTEE_ACC.as_ref()).unwrap();
+        transaction.set_signature(signature);
 
-        async fn get_block(&self, _block: Option<u64>) -> VdrResult<BlockDetails> {
-            todo!()
-        }
+        transaction
+    }
 
-        async fn get_transaction(&self, _hash: &[u8]) -> VdrResult<Option<Transaction>> {
-            todo!()
+    pub fn read_transaction() -> Transaction {
+        Transaction {
+            type_: TransactionType::Read,
+            from: None,
+            to: VALIDATOR_CONTROL_ADDRESS.clone(),
+            nonce: None,
+            chain_id: CHAIN_ID,
+            data: vec![],
+            signature: RwLock::new(None),
+            hash: None,
         }
     }
 
-    pub fn mock_client() -> LedgerClient {
-        let mut client = LedgerClient::new(
+    pub fn mock_custom_client(client: Box<dyn Client>) -> LedgerClient {
+        let mut ledger_client = LedgerClient::new(
             CHAIN_ID,
             RPC_NODE_ADDRESS,
             &contracts(),
             Some(&QuorumConfig::default()),
         )
         .unwrap();
-        client.client = Box::new(MockClient {});
-        client
+
+        ledger_client.client = client;
+        ledger_client
     }
 
     mod create {
+        use crate::validator_control::test::VALIDATOR_CONTROL_NAME;
+        use mockall::predicate::eq;
+        use rstest::rstest;
+        use serde_json::Value;
+
         use super::*;
 
         #[test]
         fn create_client_test() {
             client();
         }
+
+        #[test]
+        fn create_client_invalid_node_address() {
+            let client_err = LedgerClient::new(CHAIN_ID, "..", &contracts(), None)
+                .err()
+                .unwrap();
+
+            assert!(matches!(
+                client_err,  | VdrError::ClientNodeUnreachable { .. }
+            ));
+        }
+
+        #[rstest]
+        #[case::invalid_contract_data(&VALIDATOR_CONTROL_ADDRESS, None, Some(VALIDATOR_CONTROL_NAME), Some(""), VdrError::ContractInvalidInputData)]
+        #[case::both_contract_path_and_spec_provided(&VALIDATOR_CONTROL_ADDRESS, Some(VALIDATOR_CONTROL_PATH), Some(VALIDATOR_CONTROL_NAME), None, VdrError::ContractInvalidSpec("".to_string()))]
+        #[case::non_existent_spec_path(&VALIDATOR_CONTROL_ADDRESS, Some(""), None, None, VdrError::ContractInvalidSpec("".to_string()))]
+        #[case::empty_contract_spec(&VALIDATOR_CONTROL_ADDRESS, None, None, None, VdrError::ContractInvalidSpec("".to_string()))]
+        fn test_create_client_errors(
+            #[case] address: &Address,
+            #[case] spec_path: Option<&str>,
+            #[case] name: Option<&str>,
+            #[case] abi: Option<&str>,
+            #[case] expected_error: VdrError,
```

Comment:
I think it will be much clearer and easier to pass the config itself as parameter instead of parts